bakery/src/scripts/gdocify_book.py

```diff
@@ -3,32 +3,108 @@
 import sys
 from lxml import etree
 from pathlib import Path
+import json
 
 
-def update_doc_links(doc):
+def gen_page_slug_resolver(metadata_files):
+    """Generate a page slug resolver function"""
+    page_slugs_by_uuid = {}
+    metadata_file_by_uuid = {
+        filepath.name.split(":")[1].split("-metadata")[0]:
+            filepath for filepath in metadata_files
+    }
+
+    def _get_page_slug(page_uuid):
+        """Lookup a page slug by UUID using metadata files"""
+        cached_slug = page_slugs_by_uuid.get(page_uuid)
+        if cached_slug:
+            return cached_slug
+
+        with metadata_file_by_uuid[page_uuid].open() as json_file:
+            json_data = json.load(json_file)
+            page_slug = json_data.get("slug")
+            page_slugs_by_uuid[page_uuid] = page_slug
+
+        if not page_slug:
+            raise ValueError(f"Unable to resolve page slug for {page_uuid}")
+
+        return page_slug
+
+    return _get_page_slug
+
+
+def update_doc_links(doc, book_uuid, book_slugs_by_uuid, page_slug_resolver):
     """Modify links in doc"""
+
+    def _rex_url_builder(book, page, fragment):
+        base_url = f"http://openstax.org/books/{book}/pages/{page}"
+        if fragment:
+            return f"{base_url}#{fragment}"
+        else:
+            return base_url
+
     for node in doc.xpath(
         '//x:a[@href and starts-with(@href, "/contents/")]',
         namespaces={"x": "http://www.w3.org/1999/xhtml"}
     ):
-        # This is either an intra-book link or inter-book link (we can
-        # differentiate the latter by data-book-uuid attrib). For now we're
-        # just going to convert these to point at http://openstax.org, but
-        # later we may convert one or both of these to point at REX via
-        # http://openstax.org/books/{book_slug}/pages/{page_slug}, with
-        # current gaps being looking up book_slugs from CMS by UUID as well
-        # as having a lookup table of page UUIDs to slugs.
-        node.attrib["href"] = "http://openstax.org"
+        page_link = node.attrib["href"].split("/")[-1]
+        # Link may have fragment
+        if "#" in page_link:
+            page_uuid, page_fragment = page_link.split("#")
+        else:
+            page_uuid = page_link
+            page_fragment = None
+
+        # This is either an intra-book link or inter-book link. We can
+        # differentiate the latter by data-book-uuid attrib).
+        if node.attrib.get("data-book-uuid"):
+            external_book_uuid = node.attrib["data-book-uuid"]
+            external_book_slug = book_slugs_by_uuid[external_book_uuid]
+            external_page_slug = node.attrib["data-page-slug"]
+            node.attrib["href"] = _rex_url_builder(
```

Comment:
I wonder if this should be somewhere shared with the PDF pipelines (I assume we want REX linking behavior in that pipeline as well. It's work that's perhaps not in scope with this PR, but I thought I would bring it up.

Reply:
Hmmm...I'm not sure (re: plans for the link rewriting on the PDF path). If so, perhaps we can refactor this a bit at that time to incorporate a common utility function?